network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -115,7 +122,34 @@ struct ConnectionState<T> {
     required_services: Services,
 }
 
-impl<T: Ord> ConnectionState<T> {
+impl<T: Clone + Ord + Unpin> Future for ConnectionState<T> {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut unban_id = None;
+
+        if let Some((id, sleep)) = self.unban_timeout.as_mut() {
+            if sleep.poll_unpin(cx).is_ready() {
+                unban_id = Some(id.clone());
+
+                if let Some((id, deadline)) = self.unban_deadlines.pop_front() {
+                    self.unban_timeout = Some((id, Box::pin(sleep_until(deadline))));
```

Comment:
Optimally, this would first check whether the deadline has already expired. In this case, we could avoid creating a `sleep_until`.